moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -30,25 +29,55 @@
  * @author Jesse Wilson
  */
 abstract class ClassFactory<T> {
+
+  /**
+   * This can't be a {@link LinkedHashTreeMap} since {@link Class}
+   * does not implement {@link Comparable}.
+   */
+  private static Map<Class<?>, ClassFactory<?>> unsafeFactories = new HashMap<>();
```

Comment:
`LinkedHashMap`.
This should probably be cached somewhere more appropriate, like the Moshi instance.
But, even moreover, I don't think it actually needs to be cached. The Factory is supposed to do work to create the Adapter; the created Adapter is still fast. This is also an independent change. Revert?